### PR TITLE
Improve Nest GPU Utilization 5/N: Async H2D in Dataloader

### PR DESCRIPTION
**Pull request by MahmoudAshraf97**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

This PR closes the gap between ASR and NEST SSL gpu utilization during training
since `AudioNoiseBatch` uses a custom return type, torch dataloader `pin_memory` is a no-op and all transfers are pageable, implementing this raised the utilization on 8xH100 machine from 80% to almost 100%

This is a 1k test run using the same config before and after

gpu_util_comparison

This PR doesn't overlap with the changes in #15589 and can be merged cleanly
**Collection**: [Note which collection this PR will affect]

# Changelog

- Add specific line by line info of high level changes in this PR.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?
@stevehuang52 

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by stevehuang52** [comment]
/ok to test [9d5548b](https://github.com/NVIDIA-NeMo/Speech/pull/15920/commits/9d5548bf25d36fe2ca1b6ea03fb223606abaae1d)

**Comment by stevehuang52** [comment]
/ok to test [9d5548b](https://github.com/NVIDIA-NeMo/Speech/pull/15920/commits/9d5548bf25d36fe2ca1b6ea03fb223606abaae1d)

**Comment by stevehuang52** [comment]
/ok to test [1a6069a](https://github.com/NVIDIA-NeMo/Speech/pull/15920/commits/1a6069a30f3a4386be7ab473fabbea2a9298b64c)

**Review by stevehuang52 — APPROVED** [review]
LGTM, thanks!
